step-plans/step-plans-simple-yaml-parser/src/main/java/step/plans/simple/YamlPlanSerializer.java

```diff
@@ -0,0 +1,230 @@
+		}
+
+		SimpleYamlPlan simplePlan = yamlMapper.treeToValue(simplePlanJsonNode, SimpleYamlPlan.class);
+		return convertSimplePlanToFullPlan(simplePlan);
+	}
+
+	private String upgradeSimpleYamlIfRequired(String bufferedYamlPlan) throws JsonProcessingException {
+		if (currentVersion != null) {
+			Document simplePlanDocument = yamlMapper.readValue(bufferedYamlPlan, Document.class);
+			String planVersionString = simplePlanDocument.getString(SimpleYamlPlan.VERSION_FIELD_NAME);
+
+			// planVersionString == null means than no migration is required (version is actual)
+			if (planVersionString != null) {
+				log.info("Migrating simple plan from version {} to {}", planVersionString, currentVersion);
+
+				// convert yaml plan to document to perform migrations
+				CollectionFactory tempCollectionFactory = new InMemoryCollectionFactory(new Properties());
+				Version planVersion = new Version(planVersionString);
+				Collection<Document> tempCollection = tempCollectionFactory.getCollection(SIMPLE_PLANS_COLLECTION_NAME, Document.class);
+				Document planDocument = tempCollection.save(simplePlanDocument);
+
+				// run migrations (AbstractSimplePlanMigrationTask)
+				migrationManager.migrate(tempCollectionFactory, planVersion, currentVersion);
+
+				Document migratedDocument = tempCollection.find(Filters.id(planDocument.getId()), null, null, null, 0).findFirst().orElseThrow();
+
+				// remove automatically generated document id
+				migratedDocument.remove(AbstractIdentifiableObject.ID);
+
+				// convert document back to the yaml string
+				bufferedYamlPlan = yamlMapper.writeValueAsString(migratedDocument);
+
+				if(log.isDebugEnabled()){
+					log.debug("Simple plan after migrations: {}", bufferedYamlPlan);
+				}
+			}
+		}
+		return bufferedYamlPlan;
+	}
+
+	public ObjectMapper getYamlMapper() {
+		return yamlMapper;
+	}
+
+	private Plan convertSimplePlanToFullPlan(SimpleYamlPlan simpleYamlPlan) {
+		Plan fullPlan = new Plan(simpleYamlPlan.getRoot().getAbstractArtefact());
+		fullPlan.addAttribute("name", simpleYamlPlan.getName());
+		applyDefaultValues(fullPlan);
+		return fullPlan;
+	}
+
+	private void applyDefaultValues(Plan fullPlan) {
+		if (this.idGenerator != null) {
+			fullPlan.setId(this.idGenerator.get());
+		}
+
+		AbstractArtefact root = fullPlan.getRoot();
+		if (root != null) {
+			applyDefaultValuesForArtifact(root);
+		}
+	}
+
+	private void applyDefaultValuesForArtifact(AbstractArtefact artifact) {
+		if (this.idGenerator != null) {
+			artifact.setId(this.idGenerator.get());
+		}
+		applyDefaultValuesForChildren(artifact);
+	}
+
+	private void applyDefaultValuesForChildren(AbstractArtefact root) {
+		List<AbstractArtefact> children = root.getChildren();
+		if (children != null) {
+			for (AbstractArtefact child : children) {
+				applyDefaultValuesForArtifact(child);
+			}
+		}
+	}
+
+	/**
+	 * Write the full plan as YAML
+	 */
+	public void toFullYaml(OutputStream os, Plan plan) throws IOException {
```

Comment:
I would define what is exactly meant with "full" and "simple" plans

Reply:
@jeromecomte 
Yes, I agree. I would remove all public methods working with "full" format from this class at all. So this class will provide the pure methods to read the plan from yaml (new format) and write the plan to yaml (new format).